python/ray/client_builder.py

```diff
@@ -100,19 +107,36 @@ def connect(self) -> ClientContext:
                 includes the server's version of Python & Ray as well as the
                 dashboard_url.
         """
+        old_ray_cxt = None
+        if ray.util.client.ray.is_connected() and self._allow_multiple_cli:
+            raise RuntimeError("It's in single-client mode. We can't create"
```

Comment:
What example code reproduces this error? Seems confusing to raise an error if allow_multiple is True.

Reply:
see above comment.

Reply:
I still don't understand, can you provide a code example? Also, the message text should be improved.

Reply:
```
ray.init("ray://")

# The following will error
ray.init("ray://", allow_multiple=True)
```

Reply:
Thanks! Actually, I think in this case we should allow the connection to proceed. There's no chance of ambiguity here.

Reply:
there are two things here:
- init
  - with allow_multiple it'll initialize the non-default one
  - without this, it'll initialize the default one
  - if you call `ray.util.client.ray.is_connected` directly without `with` it'll use the default one. so if allow_multiple is set to be true, even you have a handler, it still will return false.
- shutdown
  - if you call ray.shutdown, if it's running without `with`, it'll shutdown all clients. otherwise, it'll just shutdown that context.

I think there are some API not specified clearly: 1) how to define it's connected; 2) which API should apply to all.

Reply:
Either way, I think we should allow

```
ray.init()
ray.init(allow_multiple=True)
```

The reason is this: say you are writing a library function which uses multiple client connections. Then, your library function cannot be called unless the user of your function passes allow_multiple=True, which is weird.

Make sense?

Reply:
ok. but how about shutdown and disconnect?